feat(typescript-estree)!: throw error on file not in project when `project` set

uniqueiniquity opened the pull request:
BREAKING CHANGE

Currently, when using `@typescript-eslint/parser` with the `project` option, there is no connection between the set of files as defined by the `tsconfig.json` file(s) you provide and the files that you're actually linting. It turns out that if you lint a file not in that set, the process of building enough context to drive semantic rules incurs a significant performance hit.

Here, we introduce an error when such a file is linted when the `project` setting is provided. This will help users avoid incurring this performance hit, while explicitly alerting them of what issue is occurring (rather than simply not running semantic lint rules on such files). Note that this error matches TSLint's behavior in this scenario.

Fixes #724

bradzacher commented on packages/parser/README.md:
```diff
@@ -48,7 +48,7 @@ The following additional configuration options are available by specifying them
 - **`useJSXTextNode`** - default `true`. Please set `false` if you use this parser on ESLint v4. If this is `false`, the parser creates the AST of JSX texts as the legacy style.
 
-- **`project`** - default `undefined`. This option allows you to provide a path to your project's `tsconfig.json`. **This setting is required if you want to use rules which require type information**. You may want to use this setting in tandem with the `tsconfigRootDir` option below.
+- **`project`** - default `undefined`. This option allows you to provide a path to your project's `tsconfig.json`. **This setting is required if you want to use rules which require type information**. You may want to use this setting in tandem with the `tsconfigRootDir` option below. Note that if this setting is specified, you must only lint files that are included in the projects as defined by t ...
```
WDYT?

Would it good to suggest that users create a separate `tsconfig.eslint.json` which is essentially

```cjson
{
  "extends": "./tsconfig.json", // path to base tsconfig
  "includes": [
    "src/**/*.ts",
    "test/**/*.ts",
    // etc
  ]
}
```

uniqueiniquity replied:
Good idea! Added.

bradzacher commented on packages/typescript-estree/src/parser.ts:
```diff
@@ -67,31 +64,22 @@ function resetExtra(): void {
- * @param options The config object
- * @returns If found, returns the source file corresponding to the code and the containing program
- */
-function getASTAndDefaultProject(code: string, options: TSESTreeOptions) {
```
do we want to leave this in behind an option (default false) in case there are some cases that we didn't think of?

uniqueiniquity replied:
Added the option back.

uniqueiniquity commented:
@bradzacher I've addressed your comments and synced with master.
The tests for `no-unnecessary-assertion` are totally busted on my machine but passing fine here... with or without my most recent commits. I'll probably delete the whole folder and try again locally.

I haven't had a chance to try with vue yet, but I will as soon as I can (tomorrow or this weekend).
I'm not concerned, though, as long as the vue parser is giving us typescript code to parse with the `.vue` filename, and as long as projects are set up as the [vetur documentation](https://vuejs.github.io/vetur/setup.html#project-setup) suggests.

uniqueiniquity commented:
Actually, just tried it with a project generated by the Vue CLI and hit an issue.
Will investigate more when less tired :)

bradzacher commented:
Just to confirm - did you try that on a SFC (Single File Component)? (eg below)

Based on #404 I wasn't sure if it'd work.
If so then that's great! We can mark that as closed.

Just for sanity can you add a test like this, just so we can make sure that we don't break support again.
(I've never used vue, but I'm pretty sure that this is valid vue code):

```js
// .eslint.js
module.exports = {
  parser: 'vue-eslint-parser',
  parserOptions: {
    "parser": '@typescript-eslint/parser',
    "project": `${__dirname}/tsconfig.json`,
    extraFileExtensions: ['.vue'],
  },
  plugins: ['@typescript-eslint'],
  rules: {
    '@typescript-eslint/no-implicit-any': 'error',
  },
};
```

```vue
<!-- Hello.vue -->
<template>
    <div>
        <div>
		    <!-- !!!!! expected error !!!!! -->
			Hello {{ (name as any) }}{{exclamationMarks}}
		</div>
        <button @click="decrement">-</button>
        <button @click="increment">+</button>
    </div>
</template>

<script lang="ts">
import Vue from "vue";
export default Vue.extend({
    props: ['name', 'initialEnthusiasm'],
    data() {
        return {
            enthusiasm: this.initialEnthusiasm,
        }
    },
    methods: {
        increment() { this.enthusiasm++; },
        decrement() {
            if (this.enthusiasm > 1) {
                this.enthusiasm--;
            }
        },
    },
    computed: {
        exclamationMarks(): any { // !!!!! expected error !!!!!
            return Array(this.enthusiasm + 1).join('!');
        }
    }
});
</script>
```

uniqueiniquity commented:
Yep, I can lint that file fine (assuming you meant `no-explicit-any`).
One thing to point out is that only the `any` in the `script` tag is flagged.
In order to traverse the template, the rule needs to use the `context.parserServices.defineTemplateBodyVisitor`, and since ours isn't vue-specific, we don't traverse the expression containing the other `any`.

Is the best way to add that test via another integration test with docker?
I'm happy to do it, but I don't really know anything about docker so it might take me a bit :)

bradzacher commented:
Awesome, it's good that it works!

> Is the best way to add that test via another integration test with docker?

That's a good question for @JamesHenry - I don't know that test harness at all.


---

> only the any in the script tag is flagged
> context.parserServices.defineTemplateBodyVisitor

I didn't know about that function ...
Do we want to add `defineTemplateBodyVisitor` (in a separate PR)?

JamesHenry commented:
I'll add the integration test in a follow up, thanks again @uniqueiniquity!
